### ISSUE TITLE
Take more inspiration from SPI NOR Flash commands?

### ISSUE THREAD

**Issue by mithro** [xobs/spibone#4]
SPI NOR Flash is frequently used to provide memory to embedded systems. It could thus be a good source of inspiration for how a `spibone` protocol should look like.

misoc also has a very good example of how to create a low resource mapping between wishbone read requests and the more high performance SPI flash read requests. See https://github.com/m-labs/misoc/blob/master/misoc/cores/spi_flash.py for example.

As many things people put on wishbone buses effectively work like memory. This would also means there is the potentially high level of overlap between how a SPI NOR flash operates and how you want to operate with wishbone.

**Comment by xobs** [comment]
I'm not aware of any SPI flash devices that have variable write times, though perhaps they do exist.  This is the reason for the sentinel value to indicate that data is ready.

Additionally, most SPI flash parts use 24-bit addressing, and while it's possible to extend it to 32-bits, doing so is not standard.

I had the idea of using the initial byte to do things like setting the bus width.  Right now we assume 32-bits, but I could see adapting this for 8, 16, and 64-bit bus widths.  Additionally, Etherbone supports "FIFO Mode" where the address pointer isn't incremented for subsequent operations, which could be a neat thing to add.

What advantage would an JEDEC-like interface provide?

**Comment by mithro** [comment]
There are 4-byte SPI flash commands,
```python
    # 4-byte address opcodes - used on Spansion and some Macronix flashes.
    READ_4B           = _Op(0x13, "Read data bytes (low frequency)")
    READ_FAST_4B      = _Op(0x0c, "Read data bytes (high frequency)")
    READ_1_1_2_4B     = _Op(0x3c, "Read data bytes (Dual Output SPI)")
    READ_1_2_2_4B     = _Op(0xbc, "Read data bytes (Dual I/O SPI)")
    READ_1_1_4_4B     = _Op(0x6c, "Read data bytes (Quad Output SPI)")
    READ_1_4_4_4B     = _Op(0xec, "Read data bytes (Quad I/O SPI)")
    READ_1_1_8_4B     = _Op(0x7c, "Read data bytes (Octal Output SPI)")
    READ_1_8_8_4B     = _Op(0xcc, "Read data bytes (Octal I/O SPI)")

    PP_4B             = _Op(0x12, "Page program (up to 256 bytes)")
    PP_1_1_4_4B       = _Op(0x34, "Quad page program")
    PP_1_4_4_4B       = _Op(0x3e, "Quad page program")
    PP_1_1_8_4B       = _Op(0x84, "Octal page program")
    PP_1_8_8_4B       = _Op(0x8e, "Octal page program")
```

We could have 8 / 16 / 32 bit address wishbone read / write opcodes in a similar manner?

**Comment by mithro** [comment]
Some SPI flash commands also require a number of "dummy bits" be sent between the address and data results.

**Comment by xobs** [comment]
Yes, many SPI flash commands have a fixed number of dummy bits.  The problem with Wishbone is that devices can take a variable amount of time.  For example, if you're accessing memory-mapped SPI flash with a div of 8, then it'll take at least 64 clock cycles for that word to be fetched.  Contrast that with accessing RAM, which will happen in as little as 0 cycles.
